Explains datums, atoms, areas, turfs, mobs and objs (#7)
spacestation13/DMByExample: src/objs/basic_types.md

> ## Basic Object Types

Reviewer: stick to 'basic object types' or 'primitive object types'

Author: yeah i didnt know which one to pick, since i don't know the difference between the two terms really. Which one should i go with?

Reviewer: probably primitive

In Object Oriented programming, objects inherit behaviour from parents. DM has some built in ancestors which share certain behaviours. These will be gone over individually and in more detail in their own pages, but here is the general overview.
### Datums

> The first is the Datum (`/datum/`). It's the ancestor of all the other types (except for /world, /client, /list, and /savefile) that we will see. It's essentially pure data, hence the name, and can have pretty much any vars you like. When you define a new "top level" object, if you do not specify a parent_type, it defaults to /datum.

Reviewer: ```suggestion
The first is the Datum (`/datum/`). It's the ancestor of all the other types (except for some types like /world, /client, /list, and more) that we will see. It's essentially pure data, hence the name, and can have pretty much any vars you like. When you define a new "top level" object, if you do not specify a parent_type, it defaults to /datum.
```

```dm
MyType
	var/myvar = "test"
// this "mytype object" is a datum. Normally, though, you'd put it as datum/MyType or /datum/MyType
```
### Atoms

> Atoms (`/atom/`) are the direct child of datums, and they represent objects that can go on a map and therefore in the world. It's an abstract parent which shouldn't really be used by itself, but it gives important information to its children, mainly areas, turfs, objs, and mobs.

Reviewer: ```suggestion
Atoms (`/atom/`) are the direct child of datums, and they represent objects that can go on a map and therefore in the world. It's an abstract parent which shouldn't really be used by itself, but it gives important information to its children, mainly **a**reas, **t**urfs, **o**bjs, and **m**obs.
```

Author: :o

> `/atom/movable` defines behaviour related to moving things, and is the parent of `/obj` and `/mob`. Again, this is an abstract parent, so you shouldn't define objects of its type.

Reviewer: > so you shouldn't define objects of its type.

unclear

#### Areas

> An area (`/area/`) is how the game controls rooms and certain zones. Every tile on a map has exactly one area, and if you don't set one in the map editor, it defaults to `/area/space`.

Reviewer: >  Every object on a map has exactly one area

no. you can have null loc

> if you don't set one in the map editor, it defaults to `/area/space`.

no, it's some var

> When the game starts, one object is created per area, and it's treated as a single object. All tiles with the same area are connected to the same single instance of that area.

Reviewer: > When the game starts, one object is created per area

wat? are you trying to say `/area/foo` is a singleton? this is unclear wording. Don't toss 'object' around because it's confusing

Author: ah meant to say instance. was looking at the dm docs for reference and i didn't fully get it
![image](https://github.com/spacestation13/DMByExample/assets/51260013/f211be5a-b081-4872-871a-0d050cd1a14c)

Author: ended up just removing that bit altogether. That can go in area's own page later if need be.

```dm
/area/Entered(O)
	.=..() // makes sure that the parent stuff is called when the function returns, for instance, from /atom.
	if(desc)
		O << desc // makes it so that entering any area will return the description, if there is one.
/area/outside
	desc = "What a lovely day."
/area/inside
	desc = "Wonder what the weather's like."
```
#### Turfs
A turf (`/turf/`is essentially the tile itself. They can't be moved, only replaced with new ones (which removes the old one). Depending on what game or what server you're running, these are generally used as floors, walls and windows (although, of course, it varies).

> When you do src.loc on an atom, the return value is actually the turf.

Reviewer: > When you do src.loc on an atom, the return value is actually the turf.

not always true, also weird wording

```dm
/turf/wall
	desc = "A steel wall."
	density = 1 // it can't be walked through.
/turf/floor
	desc = "A steel tiled floor"

> 	specialfloor

Reviewer: please don't poison readers with absolute pathing

		desc = "Something seems strange about this floor."
		Entered() // when the tile is walked over, do a special thing
			doSpecialThing()
		proc/doSpecialThing()
			// the thing that the tile would do when walked on would go here.
```
#### Mobs
Living things that can move around, deriving from the word "mobile". These are what the player controls, and are slightly more complicated than objs since they can have a /client attached to them (i.e. a player).
```dm
/mob/human
	desc = "A regular old human."
```
#### Objs

> Objs (NOT the same as objects) are general purpose items and things that you can find in a map. Everything that's not a turf or a mob on a map is almost certainly an obj.

Reviewer: > Objs (NOT the same as objects)

unclear.

## Basic Object Types

In Object Oriented programming, objects inherit behaviour from parents. DM has some built in ancestors which share certain behaviours. These will be gone over individually and in more detail in their own pages, but here is the general overview.

### Datums

The first is the Datum (`/datum/`). It's the ancestor of all the other types (except for some types like /world, /client, /list, and more) that we will see. It's essentially pure data, hence the name, and can have pretty much any vars you like.
> Note: When you define a new "top level" object, if you do not specify a parent_type, it defaults to /datum.

```dm
MyType
	var/myvar = "test"
// this "mytype object" is a datum. Normally, though, you'd put it as datum/MyType or /datum/MyType
```

### Atoms

Atoms (`/atom/`) are the direct child of datums, and they represent objects that can go on a map and therefore in the world. It's an abstract parent which shouldn't really be used by itself, but it gives important information to its children, mainly **a**reas, **t**urfs, **o**bjs, and **m**obs.

> `/atom/movable` defines behaviour related to moving things, and is the parent of `/obj` and `/mob`. Again, this is an abstract parent, so you shouldn't define objects under it.

Reviewer: Why not? Goon does.

Author: since when 💀 
i'll remove the abstract parent warning then

```dm
/atom/movable/car
	//don't do this
```

#### Areas

An area (`/area/`) is how the game controls rooms and certain zones. Every tile on a map should have exactly one area.

```dm
/area/Entered(O)
	.=..() // makes sure that the parent stuff is called when the function returns, for instance, from /atom.

> 	if(desc)

Reviewer: ```suggestion
	if (desc)
```

		O << desc // makes it so that entering any area will return the description, if there is one.
/area/outside
	desc = "What a lovely day."
/area/inside
	desc = "Wonder what the weather's like."
```

#### Turfs

> A turf (`/turf/`is essentially the tile itself. They can't be moved, only replaced with new ones (which removes the old one). Depending on what game or what server you're running, these are generally used as floors, walls and windows (although, of course, it varies).

Reviewer: ```suggestion
A turf (`/turf/`) is essentially the tile itself. They can't be moved, only replaced with new ones (which removes the old one). Depending on what game or what server you're running, these are generally used as floors, walls and windows (although, of course, it varies).
```

> Note: the return value of .loc on an atom is usually the turf that the atom is on. Unless its location is null, of course.

```dm
/turf/wall
	desc = "A steel wall."
	density = 1 // it can't be walked through.
/turf/floor
	desc = "A steel tiled floor"

> /turf/floorspecialfloor

Reviewer: ```suggestion
/turf/floor/special
```
